README.md
# PhoneNumberInput Flutter Widget 📱
Welcome to the initial release of PhoneNumberInput, a Flutter widget that allows users to input phone numbers with country code selection. Stay tuned for more exciting features in future releases!🚀

## Example 🌈



<table>
<tr>
<td>

```dart
PhoneNumberInput(
  labelText: 'Phone Number',
  isDarkMode: false,
  onCountryCodeChanged: (code) {
    print('Country code changed: $code');
  },
  onPhoneNumberChanged: (phoneNumber) {
    print('Phone number changed: $phoneNumber');
  },
)
```


```dart
PhoneNumberInput(
  labelText: 'Phone Number',
  isDarkMode: true,
  onCountryCodeChanged: (code) {
    print('Country code changed: $code');
  },
  onPhoneNumberChanged: (phoneNumber) {
    print('Phone number changed: $phoneNumber');
  },
)
```

</td>
<td>
<div style="display:flex;">
  <img  src="./example/assets/phoneNumber_InputDemo%20(3).jpeg" style="height: 250px;"  alt="">
  <img  src="./example/assets/phoneNumber_InputDemo%20(2).jpeg" style="height: 250px;"  alt="">
</div>
  <div style="display:flex;">
  <img  src="./example/assets/phoneNumber_InputDemo%20(1).jpeg" style="height: 250px;"  alt="">
  <img  src="./example/assets/phoneNumber_InputDemo%20(4).jpeg" style="height: 250px;"  alt="">
</div>
</td>
</tr>
</table>


## Features 🌟
Phone number input with country code selection
Customizable label, styles, and dark mode support
Callbacks for country code and phone number changes.

## Installation 🛠️
To use this widget in your Flutter project, add the following to your `pubspec.yaml` file:
```yaml
dependencies:
  phonenumber_input: ^1.0.0
```
Then, run:
```js
flutter pub get
```

## Usage 🚨

```dart
import 'package:flutter/material.dart';
import 'package:phonenumber_input/phonenumber_input.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: Text('PhoneNumberInput Demo'),
        ),
        body: Center(
          child: PhoneNumberInput(
            labelText: 'Phone Number',
            isDarkMode: false,
            onCountryCodeChanged: (code) {
              print('Country code changed: $code');
            },
            onPhoneNumberChanged: (phoneNumber) {
              print('Phone number changed: $phoneNumber');
            },
          ),
        ),
      ),
    );
  }
}
```

## Widget Properties 🎨
- `labelText (String)`: The label text for the phone number input.
- `padding (EdgeInsetsGeometry)`: Padding for the widget.
- `labelStyle (TextStyle)`: Style for the label text.
- `countryCodeStyle (TextStyle)`: Style for the country code text.
- `isDarkMode (bool)`: Set to true for dark mode, false for light mode.
- `onCountryCodeChanged (Function(String)?)`: Callback for country code changes.
- `onPhoneNumberChanged (Function(String)?)`: Callback for phone number changes.


Thank you for using PhoneNumberInput! Feel free to contribute, report issues, or suggest new features. Enjoy coding with Flutter! 🎉


CHANGELOG.md
## 0.0.1

Initial Phone Number Input with country code selector.

## 0.0.2

Update the package to support with dart sdk 3.2.1

## 0.0.3

Optimise the code for better performance

example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility in the flutter_test package. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:phonenumber_input/main.dart';

void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(const MyApp());

    // Verify that our counter starts at 0.
    expect(find.text('0'), findsOneWidget);
    expect(find.text('1'), findsNothing);

    // Tap the '+' icon and trigger a frame.
    await tester.tap(find.byIcon(Icons.add));
    await tester.pump();

    // Verify that our counter has incremented.
    expect(find.text('0'), findsNothing);
    expect(find.text('1'), findsOneWidget);
  });
}


test/phonenumber_input_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:phonenumber_input/phonenumber_input.dart';

void main() {}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:flutter/scheduler.dart';
import 'package:phonenumber_input/phonenumber_input.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      title: 'Flutter Demo',
      theme: ThemeData(
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.amber),
        useMaterial3: true,
      ),
      home: const MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key});

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  Brightness brightness =
      SchedulerBinding.instance.platformDispatcher.platformBrightness;
  bool darkModeOn = false;

  @override
  void initState() {
    super.initState();
    darkModeOn = brightness == Brightness.dark;
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor:
          darkModeOn ? const Color.fromARGB(221, 48, 48, 48) : Colors.white,
      body: SingleChildScrollView(
        padding: const EdgeInsets.all(16),
        child: Column(
          children: [
            const SizedBox(
              height: 60,
            ),
            PhoneNumberInput(
              labelText: "Phone Number",
              isDarkMode: darkModeOn,
              onCountryCodeChanged: (value) {
                print(value);
              },
              onPhoneNumberChanged: (value) {
                print(value);
              },
            ),
          ],
        ),
      ),
    );
  }
}


lib/src/countrycode_selector.dart
import 'dart:convert';

import 'package:flutter/material.dart';
import 'package:flutter/services.dart';

class CountryCodeSelector extends StatefulWidget {
  const CountryCodeSelector({
    super.key,
    required this.darkModeOn,
  });

  /// Indicates whether dark mode is enabled in the selector.
  final bool darkModeOn;

  @override
  State<CountryCodeSelector> createState() => _CountryCodeSelectorState();
}

class _CountryCodeSelectorState extends State<CountryCodeSelector> {
  List countries = [];
  List filteredCountries = [];
  final TextEditingController _searchController = TextEditingController();
  var selectedCountry = {};

  @override
  void initState() {
    super.initState();
    _loadCountriesFromJson();
  }

  /// Loads the list of countries from a JSON file.
  Future<void> _loadCountriesFromJson() async {
    String jsonString = await rootBundle
        .loadString('packages/phonenumber_input/assets/directory.json');
    final data = await json.decode(jsonString);
    setState(() {
      countries = data;
      filteredCountries = countries;
    });
  }

  /// Filters the list of countries based on the search query.
  void _searchCountry(String query) {
    List filterCountries = [];
    filterCountries.addAll(countries);
    if (query.isNotEmpty) {
      List tempCountries = [];
      for (var country in filterCountries) {
        if (country['name']
            .toString()
            .toLowerCase()
            .contains(query.toLowerCase())) {
          tempCountries.add(country);
        }
      }
      setState(() {
        filteredCountries = tempCountries;
      });
      return;
    } else {
      setState(() {
        filteredCountries = countries;
      });
    }
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: widget.darkModeOn
          ? const Color.fromARGB(221, 48, 48, 48)
          : Colors.white,
      body: CustomScrollView(
        slivers: [
          SliverAppBar(
            floating: true,
            pinned: true,
            snap: false,
            backgroundColor: widget.darkModeOn
                ? const Color.fromARGB(221, 48, 48, 48)
                : Colors.white,
            leading: IconButton(
              onPressed: () => Navigator.pop(context),
              icon: Icon(
                Icons.arrow_back_ios_rounded,
                color: widget.darkModeOn ? Colors.white : Colors.black,
              ),
            ),
            title: Text(
              "Country Code",
              style: TextStyle(
                color: widget.darkModeOn ? Colors.white : Colors.black,
                fontWeight: FontWeight.w500,
                fontSize: 16,
              ),
            ),
            centerTitle: true,
            bottom: AppBar(
              automaticallyImplyLeading: false,
              backgroundColor: widget.darkModeOn
                  ? const Color.fromARGB(221, 48, 48, 48)
                  : Colors.white,
              title: Container(
                height: 40,
                decoration: BoxDecoration(
                  borderRadius: BorderRadius.circular(4),
                ),
                child: Center(
                  child: TextFormField(
                    controller: _searchController,
                    onChanged: (value) => _searchCountry(value),
                    style: TextStyle(
                      color: widget.darkModeOn
                          ? Colors.white
                          : const Color.fromRGBO(46, 46, 46, 1),
                      fontWeight: FontWeight.w500,
                      fontSize: 14,
                    ),
                    decoration: InputDecoration(
                      hintText: "Search",
                      border: OutlineInputBorder(
                        borderRadius: BorderRadius.circular(4),
                        borderSide: BorderSide(
                          color: widget.darkModeOn
                              ? Colors.white
                              : const Color.fromRGBO(46, 46, 46, 1),
                        ),
                      ),
                      contentPadding:
                          const EdgeInsets.symmetric(horizontal: 16),
                      hintStyle: TextStyle(
                        color: widget.darkModeOn
                            ? Colors.white30
                            : const Color.fromRGBO(46, 46, 46, 1),
                        fontWeight: FontWeight.w500,
                        fontSize: 14,
                      ),
                    ),
                  ),
                ),
              ),
            ),
          ),
          SliverList(
            delegate: SliverChildBuilderDelegate(
              (context, index) => ListTile(
                onTap: () {
                  setState(() {
                    selectedCountry = filteredCountries[index];
                  });
                  Navigator.pop(context, selectedCountry);
                },
                title: Text(
                  filteredCountries[index]['name'],
                  style: TextStyle(
                    color: widget.darkModeOn ? Colors.white : Colors.black,
                    fontWeight: FontWeight.w500,
                    fontSize: 16,
                  ),
                ),
                subtitle: Text(
                  filteredCountries[index]['dial_code'],
                  style: TextStyle(
                    color: widget.darkModeOn ? Colors.white : Colors.black,
                    fontWeight: FontWeight.w500,
                    fontSize: 14,
                  ),
                ),
                trailing: Icon(
                  Icons.arrow_forward_ios_rounded,
                  color: widget.darkModeOn ? Colors.white : Colors.black,
                ),
              ),
              childCount: filteredCountries.length,
            ),
          ),
        ],
      ),
    );
  }
}


lib/src/phonenumber_input.dart
import 'package:flutter/material.dart';
import 'package:phonenumber_input/src/countrycode_selector.dart';

/// A Flutter widget for inputting a phone number with a country code selector.
class PhoneNumberInput extends StatefulWidget {
  /// Creates a new [PhoneNumberInput] widget.
  ///
  /// Parameters:
  /// - [labelText]: The label text of the widget.
  /// - [padding]: The padding of the widget.
  /// - [labelStyle]: The style of the label.
  /// - [countryCodeStyle]: The style of the country code.
  /// - [isDarkMode]: If true, the widget will use dark mode.
  /// - [onCountryCodeChanged]: Callback when the country code is changed.
  /// - [onPhoneNumberChanged]: Callback when the phone number is changed.
  const PhoneNumberInput({
    super.key,
    required this.labelText,
    this.padding,
    this.labelStyle,
    this.countryCodeStyle,
    this.isDarkMode = false,
    this.onCountryCodeChanged,
    this.onPhoneNumberChanged,
  });

  /// The label text of the widget
  final String labelText;

  /// The padding of the widget
  final EdgeInsetsGeometry? padding;

  /// The style of the label
  final TextStyle? labelStyle;

  /// The style of the country code
  final TextStyle? countryCodeStyle;

  /// If true, the widget will use dark mode
  final bool isDarkMode;

  /// The callback that is called when the country code is changed
  final Function(String)? onCountryCodeChanged;

  /// The callback that is called when the phone number is changed
  final Function(String)? onPhoneNumberChanged;

  @override
  State<PhoneNumberInput> createState() => _PhoneNumberInputState();
}

class _PhoneNumberInputState extends State<PhoneNumberInput> {
  Map<String, dynamic> selectedCountry = {
    "name": "India",
    "flag": "🇮🇳",
    "code": "IN",
    "dial_code": "+91",
  };

  @override
  Widget build(BuildContext context) {
    return SizedBox(
        height: MediaQuery.of(context).size.height * 0.15,
        width: MediaQuery.of(context).size.width,
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.start,
          children: [
            Text(
              widget.labelText,
              style: widget.labelStyle ??
                  TextStyle(
                    color: widget.isDarkMode
                        ? Colors.white
                        : const Color.fromRGBO(46, 46, 46, 1),
                    fontWeight: FontWeight.w500,
                    fontSize: 14,
                  ),
            ),
            const SizedBox(height: 4),
            Row(
              children: [
                Expanded(
                  flex: 3,
                  child: Container(
                    height: MediaQuery.of(context).size.height * 0.06,
                    decoration: BoxDecoration(
                      border: Border.all(color: Colors.grey.shade300),
                      borderRadius: const BorderRadius.only(
                        topLeft: Radius.circular(4),
                        bottomLeft: Radius.circular(4),
                      ),
                    ),
                    child: InkWell(
                      onTap: () async {
                        final result = await Navigator.push(
                          context,
                          MaterialPageRoute(
                            builder: (context) => CountryCodeSelector(
                              darkModeOn: widget.isDarkMode,
                            ),
                          ),
                        );
                        if (result != null) {
                          setState(() {
                            selectedCountry = result;
                            widget.onCountryCodeChanged
                                ?.call(selectedCountry['dial_code']);
                          });
                        }
                      },
                      child: Center(
                        child: Text(
                          "${selectedCountry['code']} ${selectedCountry['dial_code']}",
                          style: widget.countryCodeStyle ??
                              TextStyle(
                                color: widget.isDarkMode
                                    ? Colors.white
                                    : const Color.fromRGBO(46, 46, 46, 1),
                                fontWeight: FontWeight.w500,
                                fontSize: 14,
                              ),
                        ),
                      ),
                    ),
                  ),
                ),
                Expanded(
                  flex: 10,
                  child: Container(
                      height: MediaQuery.of(context).size.height * 0.06,
                      decoration: const BoxDecoration(
                        border: Border(
                          top: BorderSide(
                              color: Color.fromRGBO(224, 224, 224, 1)),
                          bottom: BorderSide(
                              color: Color.fromRGBO(224, 224, 224, 1)),
                          right: BorderSide(
                              color: Color.fromRGBO(224, 224, 224, 1)),
                        ),
                        borderRadius: BorderRadius.only(
                          topRight: Radius.circular(4),
                          bottomRight: Radius.circular(4),
                        ),
                      ),
                      child: TextFormField(
                        keyboardType: TextInputType.phone,
                        style: TextStyle(
                          color: widget.isDarkMode
                              ? Colors.white
                              : const Color.fromRGBO(46, 46, 46, 1),
                          fontWeight: FontWeight.w500,
                          fontSize: 14,
                        ),
                        onChanged: (value) =>
                            widget.onPhoneNumberChanged?.call(value),
                        decoration: const InputDecoration(
                          border: InputBorder.none,
                          hintText: "Phone Number",
                          hintStyle: TextStyle(
                            color: Color.fromRGBO(224, 224, 224, 1),
                            fontWeight: FontWeight.w500,
                            fontSize: 14,
                          ),
                          isDense: true,
                          contentPadding: EdgeInsets.symmetric(
                              horizontal: 16, vertical: 10),
                        ),
                      )),
                ),
              ],
            )
          ],
        ));
  }
}
